### PR TITLE
Add container utils

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
- Add debug_string helper functions for common containers.
- Add map_find_or_die helper function

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
Google C++ style bans overloading operators on types not owned: https://google.github.io/styleguide/cppguide.html#Operator_Overloading. But Ray codebase is smaller and this might be ok. I personally think just dumping the containers to strings is more flexible for callers. But not sure if it will be more complex. Otherwise looks good.

File: src/ray/util/container_util.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+std::string debug_string(const std::map<Ts...> &c) {
+  return _container_debug_string(c);
+}
+template <typename... Ts>
+std::string debug_string(const std::unordered_map<Ts...> &c) {
+  return _container_debug_string(c);
+}
+
+template <typename C>
+const typename C::mapped_type &map_find_or_die(const C &c,
```
Comment (scv119) [inline src/ray/util/container_util.h:78]: @iycheng i remember you are trying to introduce something similar, but I can't find it in the codebase.

File: src/ray/util/container_util.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+template <typename... Ts>
+std::string debug_string(const std::set<Ts...> &c) {
+  return _container_debug_string(c);
+}
+template <typename... Ts>
+std::string debug_string(const std::unordered_set<Ts...> &c) {
+  return _container_debug_string(c);
+}
+template <typename... Ts>
+std::string debug_string(const std::map<Ts...> &c) {
```
Comment (scv119) [inline src/ray/util/container_util.h:69]: probably need support absl containers too!

**Review by scv119 — COMMENTED** [review]
LGTM, minor comments.

File: src/ray/util/container_util.h (hunk 1 above)
Comment (jjyao) [reply src/ray/util/container_util.h:78]: Yea, I didn't find anything, but if it's already there, I'm happy to remove mine.

**Review by rkooo567 — APPROVED** [review]
Btw, why do we use the python style function name (do we have some convention around it)?

**Comment by jjyao** [comment]
> Btw, why do we use the python style function name (do we have some convention around it)?

I think the answer is that we don't have naming convention here and I'm not sure which one I should follow. Currently I'm following `asio_util.h`.

Once we decide on the naming convention, we can have a separate PR to unify them all.

File: src/ray/util/container_util_test.cc [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  {
+    std::map<int, int> m{{1, 2}, {3, 4}};
+    ASSERT_EQ(map_find_or_die(m, 1), 2);
+    ASSERT_DEATH(map_find_or_die(m, 5), "");
+  }
+}
+
+}  // namespace ray
+
+int main(int argc, char **argv) {
```
Comment (mwtian) [inline src/ray/util/container_util_test.cc:41]: Is the `main()` function necessary? The test already links against `gtest_main`.

**Review by mwtian — APPROVED** [review]

File: src/ray/util/container_util_test.cc (hunk 3 above)
Comment (jjyao) [reply src/ray/util/container_util_test.cc:41]: Good question. Seems we don't need but all of our existing tests have it.

**Comment by iycheng** [comment]
I still can't find the one I introduced. I only remember @rkooo567  asked me to add tests at that time. 
we can merge this one and later if we notice something, we just deprecate the old one and use this new one.

**Pull request merged**